scanpipe/pipes/d2d.py

```diff
@@ -387,3 +421,192 @@ def purldb_match(project, extensions, logger=None):
+    path_parts_len = len(path_parts)
+
+    for index in range(1, path_parts_len - 1):
+        current_parts = path_parts[index:]
+        current_path = "/".join(current_parts)
+
+        any_match = False
+
+        for candidate in PROSPECTIVE_JAVASCRIPT_MATCH.get(from_extension, []):
+            match_type = "path"
+
+            query = Q(
+                *[
+                    Q(path__endswith=f"{current_path}{extension}")
+                    for extension in candidate["to_related"]
+                ],
+                _connector=Q.OR,
+            )
+
+            matches = to_resources.filter(query)
+
+            if not matches:
+                continue
+
+            if len(matches) > len(current_parts) + len(candidate["to_related"]):
+                continue
+
+            if _is_minified_and_map_compiled_from_source(
+                matches,
+                from_resource,
+                minified_extension=candidate["to_minified_ext"],
+            ):
+                match_type = "js_compiled"
+
+            for match in matches:
+                any_match = True
+
+                relation = CodebaseRelation.objects.filter(
+                    from_resource=from_resource,
+                    to_resource=match,
+                    map_type=match_type,
+                )
+                if not relation.exists():
+                    pipes.make_relation(
+                        from_resource=from_resource,
+                        to_resource=match,
+                        map_type=match_type,
+                        extra_data={
+                            "path_score": f"{len(current_parts)}/{path_parts_len-1}",
+                        },
+                    )
+
+        if any_match:
+            break
+
+
+def _is_minified_and_map_compiled_from_source(
+    to_resources, from_source, minified_extension
+):
+    """Check if a minified file and its map were compiled from a source file."""
+    if not minified_extension:
+        return False
+
+    path = Path(from_source.path.lstrip("/"))
+    basename, extension = _get_basename_and_extension(path.name)
+
+    source_file_name = path.name
+    source_mapping = f"sourceMappingURL={basename}{minified_extension}.map"
+
+    minified_file = None
+    minified_map_file = None
+    for resource in to_resources:
+        if resource.path.endswith(minified_extension):
+            minified_file = resource
+        elif resource.path.endswith(f"{minified_extension}.map"):
+            minified_map_file = resource
+        if minified_file and minified_map_file:
+            if _source_mapping_in_minified(minified_file, source_mapping) and (
+                _source_content_in_map(minified_map_file, from_source)
+                or _source_in_map(minified_map_file, source_file_name)
+            ):
+                return True
+            break
+
+    return False
+
+
+def _source_mapping_in_minified(resource, source_mapping):
```

Comment:
Why mark it as "private" with underscore?

Reply:
not private just an indication that it's helper function